# Limpieza y transformación de datos

## Importación de librerías y configuraciones

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# ============================
# CONFIGURACIÓN DE PANDAS
# ============================

# Mostrar todas las columnas del DataFrame
pd.set_option('display.max_columns', None)

# Ajustar el ancho de la salida al tamaño de la ventana
pd.set_option('display.width', None)

# Mostrar el contenido completo de las celdas
pd.set_option('display.max_colwidth', None)

# Mostrar los números con dos decimales
pd.options.display.float_format = '{:,.2f}'.format

# ============================
# CONFIGURACIÓN DE GRÁFICOS
# ============================

# Tamaño por defecto de las figuras
plt.rcParams['figure.figsize'] = (10, 6)

# Estilo de los gráficos
sns.set_theme(style="whitegrid")

## Definición de rutas

In [2]:
from pathlib import Path

BASE_DIR = Path.cwd().parent
DATA_DIR = BASE_DIR / "Datos_originales"
ECOMMERCE_DIR = DATA_DIR / "ecommerce"
MARKETING_DIR = DATA_DIR / "marketing"

## Carga de datos

In [3]:
# Carga de los archivos del dataset ecommerce

customers = pd.read_csv(ECOMMERCE_DIR/"olist_customers_dataset.csv")
geolocation = pd.read_csv(ECOMMERCE_DIR/"olist_geolocation_dataset.csv")
order_items = pd.read_csv(ECOMMERCE_DIR/"olist_order_items_dataset.csv")
order_payments = pd.read_csv(ECOMMERCE_DIR/"olist_order_payments_dataset.csv")
order_reviews = pd.read_csv(ECOMMERCE_DIR/"olist_order_reviews_dataset.csv")
orders = pd.read_csv(ECOMMERCE_DIR/"olist_orders_dataset.csv")
products = pd.read_csv(ECOMMERCE_DIR/"olist_products_dataset.csv")
sellers = pd.read_csv(ECOMMERCE_DIR/"olist_sellers_dataset.csv")
product_category_name = pd.read_csv(ECOMMERCE_DIR/"product_category_name_translation.csv")

In [4]:
# Carga de los archivos del dataset marketing

marketing_qualified_leads = pd.read_csv(MARKETING_DIR / "olist_marketing_qualified_leads_dataset.csv")
closed_deals = pd.read_csv(MARKETING_DIR / "olist_closed_deals_dataset.csv")

## Crear copias para limpieza y transformación

In [5]:
customers_clean = customers.copy()
orders_clean = orders.copy()
order_items_clean = order_items.copy()
order_payments_clean = order_payments.copy()
order_reviews_clean = order_reviews.copy()
products_clean = products.copy()
sellers_clean = sellers.copy()
geolocation_clean = geolocation.copy()
product_category_translation_clean = product_category_name.copy()

marketing_qualified_leads_clean = marketing_qualified_leads.copy()
closed_deals_clean = closed_deals.copy()

## Limpieza y transformación

In [6]:
# Paso 1. Convertir todas las fechas

## Conversión de columnas de fecha en la tabla orders

columnas_fecha_orders= [
    "order_purchase_timestamp",
    "order_approved_at",
    "order_delivered_carrier_date",
    "order_delivered_customer_date",
    "order_estimated_delivery_date"
]

for columna in columnas_fecha_orders:
    orders_clean[columna] = pd.to_datetime(orders_clean[columna], errors="coerce")

In [7]:
orders_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   order_id                       99441 non-null  str           
 1   customer_id                    99441 non-null  str           
 2   order_status                   99441 non-null  str           
 3   order_purchase_timestamp       99441 non-null  datetime64[us]
 4   order_approved_at              99281 non-null  datetime64[us]
 5   order_delivered_carrier_date   97658 non-null  datetime64[us]
 6   order_delivered_customer_date  96476 non-null  datetime64[us]
 7   order_estimated_delivery_date  99441 non-null  datetime64[us]
dtypes: datetime64[us](5), str(3)
memory usage: 6.1 MB


In [8]:
## order_items

order_items_clean["shipping_limit_date"] = pd.to_datetime(order_items_clean["shipping_limit_date"], errors="coerce")

In [9]:
order_items_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
 #   Column               Non-Null Count   Dtype         
---  ------               --------------   -----         
 0   order_id             112650 non-null  str           
 1   order_item_id        112650 non-null  int64         
 2   product_id           112650 non-null  str           
 3   seller_id            112650 non-null  str           
 4   shipping_limit_date  112650 non-null  datetime64[us]
 5   price                112650 non-null  float64       
 6   freight_value        112650 non-null  float64       
dtypes: datetime64[us](1), float64(2), int64(1), str(3)
memory usage: 6.0 MB


In [10]:
## order_reviews

columnas_fecha_reviews = [
    "review_creation_date",
    "review_answer_timestamp"
]

for columna in columnas_fecha_reviews:
    order_reviews_clean[columna] = pd.to_datetime(order_reviews_clean[columna], errors="coerce")

In [11]:
order_reviews_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  str           
 1   order_id                 99224 non-null  str           
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  str           
 4   review_comment_message   40977 non-null  str           
 5   review_creation_date     99224 non-null  datetime64[us]
 6   review_answer_timestamp  99224 non-null  datetime64[us]
dtypes: datetime64[us](2), int64(1), str(4)
memory usage: 5.3 MB


In [12]:
## marketing_qualified_leads

marketing_qualified_leads_clean["first_contact_date"] = pd.to_datetime(marketing_qualified_leads_clean["first_contact_date"], errors="coerce")

In [13]:
marketing_qualified_leads_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   mql_id              8000 non-null   str           
 1   first_contact_date  8000 non-null   datetime64[us]
 2   landing_page_id     8000 non-null   str           
 3   origin              7940 non-null   str           
dtypes: datetime64[us](1), str(3)
memory usage: 250.1 KB


In [14]:
## closed_deals

closed_deals_clean["won_date"] = pd.to_datetime(closed_deals_clean["won_date"], errors="coerce")

In [15]:
closed_deals_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   mql_id                         842 non-null    str           
 1   seller_id                      842 non-null    str           
 2   sdr_id                         842 non-null    str           
 3   sr_id                          842 non-null    str           
 4   won_date                       842 non-null    datetime64[us]
 5   business_segment               841 non-null    str           
 6   lead_type                      836 non-null    str           
 7   lead_behaviour_profile         665 non-null    str           
 8   has_company                    63 non-null     object        
 9   has_gtin                       64 non-null     object        
 10  average_stock                  66 non-null     str           
 11  business_type                 

In [16]:
# Paso 2.  Revisión de tipos de datos

## Tabla closed_deals_clean

closed_deals_clean["has_company"].value_counts(dropna=False)

has_company
NaN      779
True      58
False      5
Name: count, dtype: int64

In [17]:
closed_deals_clean["has_gtin"].value_counts(dropna=False)

has_gtin
NaN      778
True      54
False     10
Name: count, dtype: int64

In [18]:
closed_deals_clean["average_stock"].value_counts(dropna=False)

average_stock
NaN        776
5-20        22
50-200      15
1-5         10
20-50        8
200+         7
unknown      4
Name: count, dtype: int64

No se van a transformar los tipos de datos de estas dos columnas de la tabla closed_deals_clean ya que con tan pocos valores informados, es muy probable que estas columnas no aporten valor al análisis.

In [19]:
# Paso 3. Revisión de valores nulos

## Tabla products

products_clean[products_clean["product_category_name"].isna()].head()

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
105,a41e356c76fab66334f36de622ecbd3a,NaN,NaN,NaN,NaN,650.00,17.00,14.00,12.00
128,d8dee61c2034d6d075997acef1870e9b,NaN,NaN,NaN,NaN,300.00,16.00,7.00,20.00
145,56139431d72cd51f19eb9f7dae4d1617,NaN,NaN,NaN,NaN,200.00,20.00,20.00,20.00
154,46b48281eb6d663ced748f324108c733,NaN,NaN,NaN,NaN,"18,500.00",41.00,30.00,41.00
197,5fb61f482620cb672f5e586bb132eae9,NaN,NaN,NaN,NaN,300.00,35.00,7.00,12.00


In [20]:
products_clean["product_id"].nunique()

32951

In [21]:
order_items_clean["product_id"].isin(products_clean[products_clean["product_category_name"].isna()]["product_id"]).sum()

np.int64(1603)

In [22]:
products_clean["product_category_name"].isna().sum()

np.int64(610)

In [23]:
products_clean["product_name_lenght"].isna().sum()

np.int64(610)

In [24]:
products_clean["product_category_name"] = products_clean["product_category_name"].fillna("Unknown")

In [25]:
products_clean["product_category_name"].isna().sum()

np.int64(0)

Tras comprobar que los productos con datos nulos en la columna product_category_name aparecen en 1603 registros en la tabla order_items (principal), se decide rellenar estos valores nulos con la categoría "Unknown" para mantener estos registros. Las demás columnas no se trataron ya que no hay ninguna base objetiva que nos permita imputar esos datos de forma confiable.

In [26]:
products_clean[products_clean["product_weight_g"].isna()]

,product_id,product_category_name,product_name_lenght,product_description_lenght,product_photos_qty,product_weight_g,product_length_cm,product_height_cm,product_width_cm
8578,09ff539a621711667c43eba6a3bd8466,bebes,60.00,865.00,3.00,NaN,NaN,NaN,NaN
18851,5eb564652db742ff8f28759cd8d2652a,Unknown,NaN,NaN,NaN,NaN,NaN,NaN,NaN


In [27]:
order_items_clean[order_items_clean["product_id"] == "09ff539a621711667c43eba6a3bd8466"]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
58833,85f8ad45e067abd694b627859fa57453,1,09ff539a621711667c43eba6a3bd8466,8b8cfc8305aa441e4239358c9f6f2485,2017-02-03 21:40:02,"1,934.00",27.00


In [28]:
order_items_clean[order_items_clean["product_id"] == "5eb564652db742ff8f28759cd8d2652a"]

,order_id,order_item_id,product_id,seller_id,shipping_limit_date,price,freight_value
7098,101157d4fae1c9fb74a00a5dee265c25,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-04-11 08:02:26,29.00,14.52
9233,1521c6bb7b1028154c8c67cf80fa809f,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-04-07 10:10:16,29.00,16.05
28715,415cfaaaa8cea49f934470548797fed1,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-04-07 10:35:19,29.00,14.52
28716,415cfaaaa8cea49f934470548797fed1,2,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-04-07 10:35:19,29.00,14.52
39299,595316a07cd3dea9db7adfcc7e247ae7,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-08-18 04:26:04,39.00,9.27
48424,6e150190fbe04c642a9cf0b80d83ee16,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-06-30 16:45:14,39.00,16.79
48980,6f497c40431d5fb0cfbd6c943dd29215,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-04-11 05:55:32,29.00,10.96
71134,a2456e7f02197951664897a94c87242d,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-04-06 11:50:09,29.00,24.84
73556,a7a43f469c0d7bdb0a23a82db125aefa,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-08-28 13:15:11,39.00,15.10
82767,bbfc7badbed2f1828e22b6d629201bd4,1,5eb564652db742ff8f28759cd8d2652a,4e922959ae960d389249c378d1c939f5,2017-07-18 04:04:10,39.00,11.85


La tabla products presentaba dos registros adicionales con valores ausentes en las variables product_weight_g, product_length_cm, product_height_cm y product_width_cm.
Tras comprobar su presencia en la tabla order_items, se verificó que ambos productos habían sido comercializados. Se decidió mantener los valores ausentes y conservar los registros evitando introducir información inventada.

In [29]:
## Tabla orders

orders_clean.loc[orders_clean["order_delivered_customer_date"].isna(), "order_status"].value_counts()

order_status
shipped        1107
canceled        619
unavailable     609
invoiced        314
processing      301
delivered         8
created           5
approved          2
Name: count, dtype: int64

In [30]:
orders_clean.loc[(orders_clean["order_status"] == "delivered") & (orders_clean["order_delivered_customer_date"].isna())]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
3002,2d1e2d5bf4dc7227b3bfebb81328c15f,ec05a6d8558c6455f0cbbd8a420ad34f,delivered,2017-11-28 17:44:07,2017-11-28 17:56:40,2017-11-30 18:12:23,NaT,2017-12-18
20618,f5dd62b788049ad9fc0526e3ad11a097,5e89028e024b381dc84a13a3570decb4,delivered,2018-06-20 06:58:43,2018-06-20 07:19:05,2018-06-25 08:05:00,NaT,2018-07-16
43834,2ebdfc4f15f23b91474edf87475f108e,29f0540231702fda0cfdee0a310f11aa,delivered,2018-07-01 17:05:11,2018-07-01 17:15:12,2018-07-03 13:57:00,NaT,2018-07-30
79263,e69f75a717d64fc5ecdfae42b2e8e086,cfda40ca8dd0a5d486a9635b611b398a,delivered,2018-07-01 22:05:55,2018-07-01 22:15:14,2018-07-03 13:57:00,NaT,2018-07-30
82868,0d3268bad9b086af767785e3f0fc0133,4f1d63d35fb7c8999853b2699f5c7649,delivered,2018-07-01 21:14:02,2018-07-01 21:29:54,2018-07-03 09:28:00,NaT,2018-07-24
92643,2d858f451373b04fb5c984a1cc2defaf,e08caf668d499a6d643dafd7c5cc498a,delivered,2017-05-25 23:22:43,2017-05-25 23:30:16,NaT,NaT,2017-06-23
97647,ab7c89dc1bf4a1ead9d6ec1ec8968a84,dd1b84a7286eb4524d52af4256c0ba24,delivered,2018-06-08 12:09:39,2018-06-08 12:36:39,2018-06-12 14:10:00,NaT,2018-06-26
98038,20edc82cf5400ce95e1afacc25798b31,28c37425f1127d887d7337f284080a0f,delivered,2018-06-27 16:09:12,2018-06-27 16:29:30,2018-07-03 19:26:00,NaT,2018-07-19


La revisión de la tabla orders mostró valores nulos en varias columnas (order_approved_at, order_delivered_carrier_date y order_delivered_customer_date). Se analizó la relación entre estos valores ausentes y el estado (order_status) de cada pedido. Se comprobó que casi todos los registros correspondían a pedidos cancelados, en proceso, no disponibles o todavía no entregados, a excepción de ocho pedidos con estado delivered que no disponían de fecha de entrega al cliente. Estos registros presentaban información consistente en el resto de variables. Dado que no existía una base objetiva para estimar la fecha de entrega real, se decidió conservar los valores nulos sin modificar la información original.

In [31]:
## Tabla order_reviews

order_reviews_clean[(order_reviews_clean["review_comment_title"].isna()) & (order_reviews_clean["review_comment_message"].isna())].shape

(56518, 7)

In [32]:
order_reviews_clean.shape

(99224, 7)

In [33]:
order_reviews_clean.loc[(order_reviews_clean["review_comment_title"].isna()) & (order_reviews_clean["review_comment_message"].isna()),
    "review_score"].value_counts().sort_index()

review_score
1     2595
2      986
3     4535
4    12864
5    35538
Name: count, dtype: int64

La tabla order_reviews presentaba un elevado número de valores ausentes en las columnas review_comment_title y review_comment_message.
Se analizó su relación con la puntuación (review_score) y se comprobó que los registros sin comentario incluían valoraciones de una a cinco estrellas, lo que indica que los clientes sí realizaron la evaluación del pedido, pero optaron por no dejar una opinión escrita. En consecuencia, no se realizaron imputaciones ni sustituciones en dichas columnas.

In [34]:
## Tabla geolocation

geolocation_clean.groupby("geolocation_zip_code_prefix").size().describe()

count   19,015.00
mean        52.60
std         72.06
min          1.00
25%         10.00
50%         29.00
75%         66.50
max      1,146.00
dtype: float64

In [35]:
geolocation_clean.groupby("geolocation_zip_code_prefix").agg({"geolocation_city": "nunique", "geolocation_state": "nunique"}).describe()

,geolocation_city,geolocation_state
count,"19,015.00","19,015.00"
mean,1.47,1.00
std,0.54,0.02
min,1.00,1.00
25%,1.00,1.00
50%,1.00,1.00
75%,2.00,1.00
max,5.00,2.00


In [36]:
# Se decide crear una tabla maestra de geolocalización que contenga: la media para los valores de latitud y longitud, y la moda para la ciudad y el país.

geolocation_master = (geolocation_clean.groupby("geolocation_zip_code_prefix").agg({
    "geolocation_lat": "mean",
    "geolocation_lng": "mean",
    "geolocation_city" : lambda x: x.mode().iloc[0],
    "geolocation_state": lambda x: x.mode().iloc[0]
}). reset_index())

In [37]:
geolocation_master.shape

(19015, 5)

In [38]:
geolocation_master.head()

,geolocation_zip_code_prefix,geolocation_lat,geolocation_lng,geolocation_city,geolocation_state
0,1001,-23.55,-46.63,sao paulo,SP
1,1002,-23.55,-46.63,sao paulo,SP
2,1003,-23.55,-46.64,sao paulo,SP
3,1004,-23.55,-46.63,sao paulo,SP
4,1005,-23.55,-46.64,sao paulo,SP


La tabla geolocation contenía más de un millón de registros, aunque únicamente existían 19.015 códigos postales distintos.
Con el objetivo de facilitar las uniones con las tablas de clientes y vendedores, se creó una tabla maestra denominada geolocation_master, compuesta por un único registro por código postal. Esta transformación permitió reducir la tabla de más de un millón de registros a 19.015 observaciones sin perder la capacidad de representar geográficamente cada código postal.

## Construcción del modelo relacional de E-commerce y dataset final

Se utiliza la tabla order_items como tabla principal ya que contiene el detalle de cada producto vendido. A partir de esta se incorporan las demás tablas para construir el dataset final

In [39]:
orders_clean["order_id"].duplicated().sum()

np.int64(0)

In [40]:
order_items_clean["order_id"].duplicated().sum()

np.int64(13984)

In [41]:
## Creación del primer dataframe de ventas 
# ## Integración de las tablas order_items y orders

df_ventas = order_items_clean.merge(orders_clean, on="order_id", how="left")

In [42]:
df_ventas.shape

(112650, 14)

In [43]:
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 14 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]
 10  order_approved_at              112635 non-null  datetime64[us]
 11  order_deliv

In [44]:
df_ventas.isna().sum()

order_id                            0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
dtype: int64

In [45]:
df_ventas.duplicated().sum()

np.int64(0)

In [46]:
df_ventas["order_id"].nunique()

98666

Como primera integración se unió order_items con orders utilizando la columna order_id como clave de relación. Se trata de una relación de uno a muchos (1:N), donde order_id es una clave primaria en orders y una clave foránea en order_items. 

In [47]:
### Integración de la tabla customers

customers_clean["customer_id"].duplicated().sum()

np.int64(0)

In [48]:
df_ventas["customer_id"].duplicated().sum()

np.int64(13984)

In [49]:
df_ventas = df_ventas.merge(customers_clean, how="left", on="customer_id")

In [50]:
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 18 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]
 10  order_approved_at              112635 non-null  datetime64[us]
 11  order_deliv

In [51]:
df_ventas.isna().sum()

order_id                            0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
dtype: int64

In [52]:
df_ventas.duplicated().sum()

np.int64(0)

La segunda integración incorporó la información de los clientes mediante la unión de customers con el dataset de ventas utilizando la columna customer_id como clave de relación. Previamente se comprobó que customer_id era una clave única en la tabla customers, mientras que en el dataset de ventas aparecía repetido, ya que un mismo cliente puede realizar varios pedidos o adquirir varios productos. La relación identificada fue de uno a muchos (1:N).

In [53]:
### Integración de la tabla products

products_clean["product_id"].duplicated().sum()

np.int64(0)

In [54]:
df_ventas["product_id"].duplicated().sum()

np.int64(79699)

In [55]:
df_ventas = df_ventas.merge(products_clean, how="left", on="product_id")

In [56]:
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 26 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]
 10  order_approved_at              112635 non-null  datetime64[us]
 11  order_deliv

In [57]:
df_ventas.isna().sum()

order_id                            0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
product_category_name               0
product_name_lenght              1603
product_description_lenght       1603
product_photos_qty               1603
product_weight_g                   18
product_length_cm                  18
product_height_cm                  18
product_width_cm                   18
dtype: int64

In [58]:
df_ventas.duplicated().sum()

np.int64(0)

 Se incorpora la información de los productos mediante la columna product_id. Esta columna actúa como clave primaria en la tabla products y como clave foránea en el dataset de ventas, estableciendo una relación de uno a muchos (1:N).

In [59]:
### Para finalizar con el bloque de productos, se une la tabla product_category_translation

df_ventas = df_ventas.merge(product_category_translation_clean, how="left", on="product_category_name")

In [60]:
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 27 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]
 10  order_approved_at              112635 non-null  datetime64[us]
 11  order_deliv

In [61]:
df_ventas.isna().sum()

order_id                            0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
product_category_name               0
product_name_lenght              1603
product_description_lenght       1603
product_photos_qty               1603
product_weight_g                   18
product_length_cm                  18
product_height_cm                  18
product_width_cm                   18
product_cate

In [62]:
df_ventas.duplicated().sum()

np.int64(0)

In [63]:
df_ventas.loc[df_ventas["product_category_name_english"].isna(), "product_category_name"].value_counts(dropna=False)

product_category_name
Unknown                                          1603
portateis_cozinha_e_preparadores_de_alimentos      15
pc_gamer                                            9
Name: count, dtype: int64

In [64]:
product_category_translation_clean[product_category_translation_clean["product_category_name"].isin(["pc_gamer","portateis_cozinha_e_preparadores_de_alimentos"])]

,product_category_name,product_category_name_english


In [65]:
df_ventas["product_category_name_english"] = df_ventas["product_category_name_english"].fillna(df_ventas["product_category_name"].map({
    "pc_gamer" : "gaming_pc",
    "portateis_cozinha_e_preparadores_de_alimentos": "portable_kitchen_and_food_preparers"})
)

Para completar la información de los productos, se integró la tabla de traducción de categorías utilizando la variable product_category_name. Se trata de una relación de uno a muchos (1:N), donde cada categoría posee una única traducción. 
Tras incorporar la tabla de traducción se identificaron dos categorías (pc_gamer y portateis_cozinha_e_preparadores_de_alimentos) que no estaban incluidas en el diccionario de traducción original, lo que generaba 24 valores nulos adicionales en la variable traducida.
Dado que estas categorías estaban correctamente identificadas en la tabla de productos y el problema se debía únicamente a la ausencia de su traducción, se decidió completar manualmente dichos valores. 

In [66]:
### Integración de la tabla sellers

sellers_clean["seller_id"].duplicated().sum()

np.int64(0)

In [67]:
df_ventas["seller_id"].duplicated().sum()

np.int64(109555)

In [68]:
df_ventas = df_ventas.merge(sellers_clean, how="left", on="seller_id")

In [69]:
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 30 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]
 10  order_approved_at              112635 non-null  datetime64[us]
 11  order_deliv

In [70]:
df_ventas.isna().sum()

order_id                            0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
product_category_name               0
product_name_lenght              1603
product_description_lenght       1603
product_photos_qty               1603
product_weight_g                   18
product_length_cm                  18
product_height_cm                  18
product_width_cm                   18
product_cate

In [71]:
df_ventas.duplicated().sum()

np.int64(0)

Una vez completada la información relativa a los productos, se inició la integración de los datos de los vendedores mediante la columna seller_id. Cada vendedor aparece una única vez en la tabla sellers, mientras que puede estar asociado a múltiples líneas de pedido dentro del dataset de ventas, confirmando una relación de uno a muchos (1:N).

In [72]:
### Tabla geolocation master

df_ventas = df_ventas.merge(geolocation_master, how="left", left_on="seller_zip_code_prefix", right_on="geolocation_zip_code_prefix")

In [73]:
(df_ventas["seller_city"] == df_ventas["geolocation_city"]).value_counts()

True     109798
False      2852
Name: count, dtype: int64

In [74]:
(df_ventas["seller_state"] == df_ventas["geolocation_state"]).value_counts()

True     111751
False       899
Name: count, dtype: int64

In [75]:
df_ventas.loc[df_ventas["seller_city"] != df_ventas["geolocation_city"],
              ["seller_zip_code_prefix",
               "seller_city",
               "geolocation_city"]
               ].drop_duplicates().head(20)

,seller_zip_code_prefix,seller_city,geolocation_city
120,14110,ribeirao preto,bonfim paulista
159,2285,sao paulo,NaN
227,14085,riberao preto,ribeirao preto
277,72460,gama,brasilia
361,4776,sp,sao paulo
432,3194,sao paulo,são paulo
455,76900,ji parana,ji-parana
460,9721,sao bernardo do capo,sao bernardo do campo
518,87140,paincandu,paicandu
691,13450,santa barbara d´oeste,santa barbara d'oeste


In [76]:
df_ventas.loc[
    df_ventas["seller_state"] != df_ventas["geolocation_state"],
    ["seller_zip_code_prefix",
        "seller_state",
        "geolocation_state"]
].drop_duplicates().head(20)

,seller_zip_code_prefix,seller_state,geolocation_state
0,27277,SP,RJ
89,37795,SP,MG
159,2285,SP,NaN
789,88301,SP,SC
1255,37708,MG,NaN
1618,81020,SP,PR
1655,31570,SP,MG
2001,71551,DF,NaN
4422,88075,SP,SC
4661,82040,PR,NaN


Se incorporó la información geográfica de los vendedores utilizando la tabla geolocation_master, creada previamente durante la fase de limpieza. La unión se realizó mediante las columnas seller_zip_code_prefix y geolocation_zip_code_prefix, ya que ambas representan el mismo identificador geográfico con distinta denominación. 
Tras incorporar la información geográfica de los vendedores se compararon las variables de ciudad y estado procedentes de la tabla sellers con las obtenidas a partir de geolocation_master. La mayoría de los registros presentaban coincidencia entre ambas fuentes; sin embargo, se identificaron algunas discrepancias debidas principalmente a diferencias de escritura, uso de abreviaturas, presencia de caracteres especiales, municipios asociados al mismo código postal y algunos códigos postales sin correspondencia en la tabla de geolocalización. Dado que no era posible determinar de forma objetiva cuál de las dos fuentes era correcta en todos los casos, se decidió conservar ambas variables en el dataset final.

FINALIZADO EL DATAFRAME DE VENTAS

Tras evaluar la posibilidad de incorporar también las tablas order_payments y order_reviews, se decidió mantenerlas como tablas independientes debido a que su nivel de granularidad es diferente. Mientras que order_items almacena una fila por producto vendido, order_reviews registra una valoración por pedido y order_payments puede contener varios registros para un mismo pedido cuando este se ha abonado mediante varios métodos de pago.
Integrar estas tablas directamente en el dataset de ventas habría provocado la repetición de información y la alteración de la granularidad de los datos, generando resultados incorrectos en agregaciones y análisis posteriores. Por este motivo, ambas tablas se conservaron independientes y se utilizarán mediante uniones puntuales únicamente cuando una pregunta de análisis requiera combinar información de distintos niveles del modelo.

In [77]:
df_ventas.shape

(112650, 35)

In [78]:
df_ventas.info()

<class 'pandas.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 35 columns):
 #   Column                         Non-Null Count   Dtype         
---  ------                         --------------   -----         
 0   order_id                       112650 non-null  str           
 1   order_item_id                  112650 non-null  int64         
 2   product_id                     112650 non-null  str           
 3   seller_id                      112650 non-null  str           
 4   shipping_limit_date            112650 non-null  datetime64[us]
 5   price                          112650 non-null  float64       
 6   freight_value                  112650 non-null  float64       
 7   customer_id                    112650 non-null  str           
 8   order_status                   112650 non-null  str           
 9   order_purchase_timestamp       112650 non-null  datetime64[us]
 10  order_approved_at              112635 non-null  datetime64[us]
 11  order_deliv

In [79]:
df_ventas.isna().sum()

order_id                            0
order_item_id                       0
product_id                          0
seller_id                           0
shipping_limit_date                 0
price                               0
freight_value                       0
customer_id                         0
order_status                        0
order_purchase_timestamp            0
order_approved_at                  15
order_delivered_carrier_date     1194
order_delivered_customer_date    2454
order_estimated_delivery_date       0
customer_unique_id                  0
customer_zip_code_prefix            0
customer_city                       0
customer_state                      0
product_category_name               0
product_name_lenght              1603
product_description_lenght       1603
product_photos_qty               1603
product_weight_g                   18
product_length_cm                  18
product_height_cm                  18
product_width_cm                   18
product_cate

In [80]:
df_ventas.duplicated().sum()

np.int64(0)

En este punto del proyecto se ha realizado una limpieza y transformación general del dataset. En la siguiente etapa ya se realizará una revisión enfocado al objetivo del proyecto, valorando las columnas que realmente nos dan información del negocio y que responden a las necesidades del proyecto que se va a presentar.

### Revisión de df_ventas según necesidades del proyecto

In [81]:
### Se eliminan columnas que no responden a las necesidades del proyecto

# Primero de la información de pedidos

df_ventas = df_ventas.drop(
    columns=["order_approved_at", "order_delivered_carrier_date", "shipping_limit_date"]
)

In [82]:
# De la información de productos

df_ventas = df_ventas.drop(
    columns=[
    "product_category_name",
    "product_name_lenght",
    "product_description_lenght",
    "product_photos_qty",
    "product_weight_g",
    "product_length_cm",
    "product_height_cm",
    "product_width_cm"]
    )

In [83]:
df_ventas.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id', 'price',
       'freight_value', 'customer_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_unique_id',
       'customer_zip_code_prefix', 'customer_city', 'customer_state',
       'product_category_name_english', 'seller_zip_code_prefix',
       'seller_city', 'seller_state', 'geolocation_zip_code_prefix',
       'geolocation_lat', 'geolocation_lng', 'geolocation_city',
       'geolocation_state'],
      dtype='str')

In [84]:
# Información de geolocalización

df_ventas = df_ventas.drop(
    columns=[
    "customer_zip_code_prefix",
    "seller_zip_code_prefix",
    "geolocation_zip_code_prefix",
    "geolocation_city",
    "geolocation_state"]
    )

In [85]:
df_ventas.columns

Index(['order_id', 'order_item_id', 'product_id', 'seller_id', 'price',
       'freight_value', 'customer_id', 'order_status',
       'order_purchase_timestamp', 'order_delivered_customer_date',
       'order_estimated_delivery_date', 'customer_unique_id', 'customer_city',
       'customer_state', 'product_category_name_english', 'seller_city',
       'seller_state', 'geolocation_lat', 'geolocation_lng'],
      dtype='str')

In [86]:
df_ventas.shape

(112650, 19)

Tras esta selección, el dataset final quedó compuesto exclusivamente por 19 variables necesarias para responder a las preguntas de negocio planteadas durante el EDA. 

In [88]:
### Se guarda el dataset final

df_ventas.to_csv("../Datos_transformados/df_ventas.csv", index=False)

Tras la decisión de no integrar las tablas order_payment y order_reviews al df_ventas, se revisa cada una de ellas para dejarlas como datasets independientes.

In [89]:
### Dataset de order_payments

order_payments_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
 #   Column                Non-Null Count   Dtype  
---  ------                --------------   -----  
 0   order_id              103886 non-null  str    
 1   payment_sequential    103886 non-null  int64  
 2   payment_type          103886 non-null  str    
 3   payment_installments  103886 non-null  int64  
 4   payment_value         103886 non-null  float64
dtypes: float64(1), int64(2), str(2)
memory usage: 4.0 MB


In [90]:
order_payments_clean.describe(include="all")

,order_id,payment_sequential,payment_type,payment_installments,payment_value
count,103886,"103,886.00",103886,"103,886.00","103,886.00"
unique,99440,NaN,5,NaN,NaN
top,fa65dad1b0e818e3ccc5cb0e39231352,NaN,credit_card,NaN,NaN
freq,29,NaN,76795,NaN,NaN
mean,NaN,1.09,NaN,2.85,154.10
std,NaN,0.71,NaN,2.69,217.49
min,NaN,1.00,NaN,0.00,0.00
25%,NaN,1.00,NaN,1.00,56.79
50%,NaN,1.00,NaN,1.00,100.00
75%,NaN,1.00,NaN,4.00,171.84


In [91]:
order_payments_clean[order_payments_clean["payment_installments"] == 0]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
46982,744bade1fcf9ff3f31d860ace076d422,2,credit_card,0,58.69
79014,1a57108394169c0b47d8f876acc9ba2d,2,credit_card,0,129.94


In [93]:
order_payments_clean[order_payments_clean["payment_value"] == 0]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
19922,8bcbe01d44d147f901cd3192671144db,4,voucher,1,0.00
36822,fa65dad1b0e818e3ccc5cb0e39231352,14,voucher,1,0.00
43744,6ccb433e00daae1283ccc956189c82ae,4,voucher,1,0.00
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.00
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.00
62674,45ed6e85398a87c253db47c2d9f48216,3,voucher,1,0.00
77885,fa65dad1b0e818e3ccc5cb0e39231352,13,voucher,1,0.00
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.00
100766,b23878b3e8eb4d25a158f57d96331b18,4,voucher,1,0.00


In [94]:
order_payments_clean[order_payments_clean["order_id"].isin([
        "4637ca194b6387e2d538dc89b124b0ee",
        "00b1cb0320190ca0daa2c88b35206009",
        "c8c528189310eaa44a745b8d9d26908b"])
]

,order_id,payment_sequential,payment_type,payment_installments,payment_value
51280,4637ca194b6387e2d538dc89b124b0ee,1,not_defined,1,0.00
57411,00b1cb0320190ca0daa2c88b35206009,1,not_defined,1,0.00
94427,c8c528189310eaa44a745b8d9d26908b,1,not_defined,1,0.00


Se identificaron tres registros con payment_type = "not_defined" y payment_value = 0. Al no existir información adicional que permitiera determinar si se trataba de registros erróneos o casos excepcionales, se decidió conservarlos sin modificaciones por representar una proporción mínima del conjunto de datos.

In [96]:
order_payments_clean["payment_type"].value_counts(dropna=False)

payment_type
credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: count, dtype: int64

Se elimina la columna sequential ya que no aporta valor.

In [97]:
order_payments_clean = order_payments_clean.drop(columns="payment_sequential")

La tabla de pagos se mantuvo independiente del dataset principal de ventas debido a que su nivel de granularidad es diferente. 
Durante la revisión se comprobó que los tipos de datos eran correctos y que no existían valores nulos. Se identificaron algunos valores atípicos, como dos registros con payment_installments = 0 y nueve registros con payment_value = 0. Dado que estos casos representan una proporción mínima del conjunto de datos y no existía evidencia suficiente para considerarlos errores, se decidió conservarlos sin modificaciones.
La única transformación aplicada consistió en eliminar la columna payment_sequential, ya que únicamente indica el orden en que se registraron los pagos de un mismo pedido y no aporta información relevante.

In [109]:
### Se guarda el dataset final

order_payments_clean.to_csv("../Datos_transformados/order_payments_clean.csv", index=False)

In [99]:
### Dataset de order_reviews

order_reviews_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 7 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   review_id                99224 non-null  str           
 1   order_id                 99224 non-null  str           
 2   review_score             99224 non-null  int64         
 3   review_comment_title     11568 non-null  str           
 4   review_comment_message   40977 non-null  str           
 5   review_creation_date     99224 non-null  datetime64[us]
 6   review_answer_timestamp  99224 non-null  datetime64[us]
dtypes: datetime64[us](2), int64(1), str(4)
memory usage: 5.3 MB


In [100]:
order_reviews_clean.describe(include="all")

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
count,99224,99224,"99,224.00",11568,40977,99224,99224
unique,98410,98673,NaN,4527,36159,NaN,NaN
top,c444278834184f72b1484dfe47de7f97,c88b1d1b157a9999ce368f218a407141,NaN,Recomendo,Muito bom,NaN,NaN
freq,3,3,NaN,423,230,NaN,NaN
mean,NaN,NaN,4.09,NaN,NaN,2018-01-12 20:49:23.948238,2018-01-16 00:23:56.977939
min,NaN,NaN,1.00,NaN,NaN,2016-10-02 00:00:00,2016-10-07 18:32:28
25%,NaN,NaN,4.00,NaN,NaN,2017-09-23 00:00:00,2017-09-27 01:53:27.250000
50%,NaN,NaN,5.00,NaN,NaN,2018-02-02 00:00:00,2018-02-04 22:41:47.500000
75%,NaN,NaN,5.00,NaN,NaN,2018-05-16 00:00:00,2018-05-20 12:11:21.500000
max,NaN,NaN,5.00,NaN,NaN,2018-08-31 00:00:00,2018-10-29 12:27:35


En describe aparece que hay 99224 reviews pero solo 98410 son únicas.

In [101]:
order_reviews_clean["review_id"].duplicated().sum()

np.int64(814)

In [102]:
order_reviews_clean[order_reviews_clean["review_id"].duplicated(keep=False)].sort_values("review_id")

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecido pela impressora, consequentemente não funcionou. Além de ter chegado com atraso de mais de 15 dias do previsto. Preciso que seja trocado.",2018-03-07,2018-03-20 18:08:23
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecido pela impressora, consequentemente não funcionou. Além de ter chegado com atraso de mais de 15 dias do previsto. Preciso que seja trocado.",2018-03-07,2018-03-20 18:08:23
90677,0115633a9c298b6a98bcbe4eee75345f,78a4201f58af3463bdab842eea4bc801,5,NaN,NaN,2017-09-21,2017-09-26 03:27:47
63193,0115633a9c298b6a98bcbe4eee75345f,0c9850b2c179c1ef60d2855e2751d1fa,5,NaN,NaN,2017-09-21,2017-09-26 03:27:47
92876,0174caf0ee5964646040cd94e15ac95e,f93a732712407c02dce5dd5088d0f47b,1,NaN,Produto entregue dentro de embalagem do fornecedor sem os parafusos de fixação das partes.,2018-03-07,2018-03-08 03:00:53
...,...,...,...,...,...,...,...
31120,fe5c833752953fed3209646f1f63b53c,4863e15fa53273cc7219c58f5ffda4fb,1,NaN,"Comprei dois produtos e ambos, mesmo enviados em dias diferentes, estão como em ""dificuldade na entrega"" segundo o rastreamento dos Correios. O que nunca me aconteceu antes.",2018-02-28,2018-02-28 13:57:52
7870,ff2fc9e68f8aabfbe18d710b83aabd30,2da58e0a7dcfa4ce1e00fad9d03ca3b5,2,NaN,NaN,2018-03-17,2018-03-19 11:44:15
82521,ff2fc9e68f8aabfbe18d710b83aabd30,1078d496cc6ab9a8e6f2be77abf5091b,2,NaN,NaN,2018-03-17,2018-03-19 11:44:15
73951,ffb8cff872a625632ac983eb1f88843c,c44883fc2529b4aa03ca90e7e09d95b6,3,NaN,NaN,2017-07-22,2017-07-26 13:41:07


In [103]:
order_reviews_clean[
    order_reviews_clean["review_id"] == "00130cbe1f9d422698c812ed8ded1919"
]

,review_id,order_id,review_score,review_comment_title,review_comment_message,review_creation_date,review_answer_timestamp
29841,00130cbe1f9d422698c812ed8ded1919,04a28263e085d399c97ae49e0b477efa,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecido pela impressora, consequentemente não funcionou. Além de ter chegado com atraso de mais de 15 dias do previsto. Preciso que seja trocado.",2018-03-07,2018-03-20 18:08:23
46678,00130cbe1f9d422698c812ed8ded1919,dfcdfc43867d1c1381bfaf62d6b9c195,1,NaN,"O cartucho ""original HP"" 60XL não é reconhecido pela impressora, consequentemente não funcionou. Além de ter chegado com atraso de mais de 15 dias do previsto. Preciso que seja trocado.",2018-03-07,2018-03-20 18:08:23


In [ ]:
### Se comprueba si estos order id con reviews duplicadas existen en orders

orders_clean[
    orders_clean["order_id"].isin([
        "04a28263e085d399c97ae49e0b477efa",
        "dfcdfc43867d1c1381bfaf62d6b9c195"
    ])
]

,order_id,customer_id,order_status,order_purchase_timestamp,order_approved_at,order_delivered_carrier_date,order_delivered_customer_date,order_estimated_delivery_date
28241,04a28263e085d399c97ae49e0b477efa,fef2e5e63da9f3e1dd89e8e319468657,delivered,2018-02-02 18:01:10,2018-02-02 18:31:27,2018-02-21 02:38:29,2018-03-11 11:32:24,2018-03-05
74048,dfcdfc43867d1c1381bfaf62d6b9c195,a7026133ddbd2e86c83ecd4dfa4dbe01,delivered,2018-02-02 18:01:08,2018-02-02 18:31:17,2018-02-05 23:46:29,2018-02-26 18:18:45,2018-03-09


Durante la revisión se detectó que algunos review_id aparecen asociados a más de un order_id. Tras comprobar los pedidos correspondientes, se observó que se trataba de pedidos distintos realizados prácticamente de forma simultánea y con la misma valoración registrada. Dado que representan pedidos diferentes y no registros duplicados completos, se decidió conservarlos sin modificaciones.

In [ ]:
### Se eliminan columnas que no aportan valor para este proyecto

order_reviews_clean = order_reviews_clean.drop(
    columns=[
        "review_comment_title",
        "review_comment_message",
        "review_answer_timestamp"]
)

In [106]:
order_reviews_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 99224 entries, 0 to 99223
Data columns (total 4 columns):
 #   Column                Non-Null Count  Dtype         
---  ------                --------------  -----         
 0   review_id             99224 non-null  str           
 1   order_id              99224 non-null  str           
 2   review_score          99224 non-null  int64         
 3   review_creation_date  99224 non-null  datetime64[us]
dtypes: datetime64[us](1), int64(1), str(2)
memory usage: 3.0 MB


In [107]:
order_reviews_clean.isna().sum()

review_id               0
order_id                0
review_score            0
review_creation_date    0
dtype: int64

In [108]:
order_reviews_clean.duplicated().sum()

np.int64(0)

In [110]:
### Se guarda el dataset final

order_reviews_clean.to_csv("../Datos_transformados/order_reviews_clean.csv", index=False)

## Dataset de marketing

In [ ]:
### Tabla marketing_qualified_leads

marketing_qualified_leads_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 4 columns):
 #   Column              Non-Null Count  Dtype         
---  ------              --------------  -----         
 0   mql_id              8000 non-null   str           
 1   first_contact_date  8000 non-null   datetime64[us]
 2   landing_page_id     8000 non-null   str           
 3   origin              7940 non-null   str           
dtypes: datetime64[us](1), str(3)
memory usage: 250.1 KB


In [116]:
marketing_qualified_leads_clean.describe(include="all")

,mql_id,first_contact_date,landing_page_id,origin
count,8000,8000,8000,7940
unique,8000,NaN,495,10
top,dac32acd4db4c29c230538b72f8dd87d,NaN,b76ef37428e6799c421989521c0e5077,organic_search
freq,1,NaN,912,2296
mean,NaN,2018-02-05 14:19:51.600000,NaN,NaN
min,NaN,2017-06-14 00:00:00,NaN,NaN
25%,NaN,2017-12-31 00:00:00,NaN,NaN
50%,NaN,2018-02-25 00:00:00,NaN,NaN
75%,NaN,2018-04-15 06:00:00,NaN,NaN
max,NaN,2018-05-31 00:00:00,NaN,NaN


In [117]:
marketing_qualified_leads_clean["mql_id"].duplicated().sum()

np.int64(0)

In [118]:
### Se elimina la columna que da información de la landing page ya que no aporta información para este proyecto en específico.

marketing_qualified_leads_clean = marketing_qualified_leads_clean.drop(columns="landing_page_id")

In [119]:
### Tabla closed_deals

closed_deals_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 14 columns):
 #   Column                         Non-Null Count  Dtype         
---  ------                         --------------  -----         
 0   mql_id                         842 non-null    str           
 1   seller_id                      842 non-null    str           
 2   sdr_id                         842 non-null    str           
 3   sr_id                          842 non-null    str           
 4   won_date                       842 non-null    datetime64[us]
 5   business_segment               841 non-null    str           
 6   lead_type                      836 non-null    str           
 7   lead_behaviour_profile         665 non-null    str           
 8   has_company                    63 non-null     object        
 9   has_gtin                       64 non-null     object        
 10  average_stock                  66 non-null     str           
 11  business_type                 

In [120]:
closed_deals_clean.describe(include="all")

,mql_id,seller_id,sdr_id,sr_id,won_date,business_segment,lead_type,lead_behaviour_profile,has_company,has_gtin,average_stock,business_type,declared_product_catalog_size,declared_monthly_revenue
count,842,842,842,842,842,841,836,665,63,64,66,832,69.00,842.00
unique,842,842,32,22,NaN,33,8,9,2,2,6,3,NaN,NaN
top,5420aad7fec3549a85876ba1c529bd84,2c43fb513632d29b3b58df74816f1b06,4b339f9567d060bcea4f5136b9f5949e,4ef15afb4b2723d8f3d81e51ec7afefe,NaN,home_decor,online_medium,cat,True,True,5-20,reseller,NaN,NaN
freq,1,1,140,133,NaN,105,332,407,58,54,22,587,NaN,NaN
mean,NaN,NaN,NaN,NaN,2018-04-23 06:04:00.891924,NaN,NaN,NaN,NaN,NaN,NaN,NaN,233.03,"73,377.68"
min,NaN,NaN,NaN,NaN,2017-12-05 02:00:00,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1.00,0.00
25%,NaN,NaN,NaN,NaN,2018-03-07 12:48:36.250000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,30.00,0.00
50%,NaN,NaN,NaN,NaN,2018-04-13 21:22:15,NaN,NaN,NaN,NaN,NaN,NaN,NaN,100.00,0.00
75%,NaN,NaN,NaN,NaN,2018-05-21 17:38:27.500000,NaN,NaN,NaN,NaN,NaN,NaN,NaN,300.00,0.00
max,NaN,NaN,NaN,NaN,2018-11-14 18:04:19,NaN,NaN,NaN,NaN,NaN,NaN,NaN,"2,000.00","50,000,000.00"


In [121]:
closed_deals_clean["mql_id"].duplicated().sum()

np.int64(0)

In [122]:
(closed_deals_clean["declared_monthly_revenue"] == 0).sum()

np.int64(797)

In [123]:
closed_deals_clean["declared_monthly_revenue"].describe()

count          842.00
mean        73,377.68
std      1,744,799.18
min              0.00
25%              0.00
50%              0.00
75%              0.00
max     50,000,000.00
Name: declared_monthly_revenue, dtype: float64

Se analizan las variables de la tabla y se decide eliminar aquellas que por falta de datos o porque no responden al objetivo del proyecto, no aportan valor.

In [124]:
closed_deals_clean = closed_deals_clean.drop(
    columns=[
        "sdr_id",
        "sr_id",
        "has_company",
        "has_gtin",
        "average_stock",
        "declared_product_catalog_size",
        "declared_monthly_revenue"]
)

In [125]:
closed_deals_clean.info()

<class 'pandas.DataFrame'>
RangeIndex: 842 entries, 0 to 841
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   mql_id                  842 non-null    str           
 1   seller_id               842 non-null    str           
 2   won_date                842 non-null    datetime64[us]
 3   business_segment        841 non-null    str           
 4   lead_type               836 non-null    str           
 5   lead_behaviour_profile  665 non-null    str           
 6   business_type           832 non-null    str           
dtypes: datetime64[us](1), str(6)
memory usage: 46.2 KB


In [126]:
closed_deals_clean.isna().sum()

mql_id                      0
seller_id                   0
won_date                    0
business_segment            1
lead_type                   6
lead_behaviour_profile    177
business_type              10
dtype: int64

In [127]:
closed_deals_clean.duplicated().sum()

np.int64(0)

Ahora unimos ambas tablas en un único dataset para el bloque de marketing

In [128]:
df_marketing = marketing_qualified_leads_clean.merge(closed_deals_clean, how="left", on="mql_id")

In [129]:
df_marketing.info()

<class 'pandas.DataFrame'>
RangeIndex: 8000 entries, 0 to 7999
Data columns (total 9 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   mql_id                  8000 non-null   str           
 1   first_contact_date      8000 non-null   datetime64[us]
 2   origin                  7940 non-null   str           
 3   seller_id               842 non-null    str           
 4   won_date                842 non-null    datetime64[us]
 5   business_segment        841 non-null    str           
 6   lead_type               836 non-null    str           
 7   lead_behaviour_profile  665 non-null    str           
 8   business_type           832 non-null    str           
dtypes: datetime64[us](2), str(7)
memory usage: 562.6 KB


In [130]:
df_marketing.isna().sum()

mql_id                       0
first_contact_date           0
origin                      60
seller_id                 7158
won_date                  7158
business_segment          7159
lead_type                 7164
lead_behaviour_profile    7335
business_type             7168
dtype: int64

In [131]:
df_marketing.duplicated().sum()

np.int64(0)

In [133]:
marketing_qualified_leads_clean["mql_id"].duplicated().sum()

np.int64(0)

In [134]:
closed_deals_clean["mql_id"].duplicated().sum()

np.int64(0)

In [135]:
df_marketing["mql_id"].duplicated().sum()

np.int64(0)

In [136]:
### Se guarda el dataset final

df_marketing.to_csv("../Datos_transformados/df_marketing.csv", index=False)

Una vez finalizada la limpieza individual de las tablas marketing_qualified_leads y closed_deals, ambas se integraron en un único conjunto de datos denominado df_marketing. La unión se realizó mediante la clave mql_id, que identifica de forma única cada lead cualificado por marketing. 